src/DeviceDataStreamValues.tsx

```diff
@@ -0,0 +1,326 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import React, { useEffect, useState } from 'react';
+import { useParams } from 'react-router-dom';
+import { Card, Container, Table } from 'react-bootstrap';
+import {
+  AstarteDataTuple,
+  AstarteDataValue,
+  AstarteDatastreamIndividualData,
+  AstarteDatastreamObjectData,
+  AstarteInterface,
+  AstarteInterfaceValues,
+} from 'astarte-client';
+import _ from 'lodash';
+
+import BackButton from './ui/BackButton';
+import { useAstarte } from './AstarteManager';
+import { AlertsBanner, useAlerts } from 'AlertManager';
+import 'react-datepicker/dist/react-datepicker.css';
+import FiltersForm from 'components/FiltersForm';
+
+function formatAstarteData(data?: AstarteDataTuple): string {
+  if (data == null) {
+    return '';
+  }
+  if (_.isArray(data)) {
+    return JSON.stringify(data);
+  }
+  if (_.isBoolean(data)) {
+    return data ? 'true' : 'false';
+  }
+  if (_.isNumber(data)) {
+    return data.toString();
+  }
+  if (_.isNull(data)) {
+    return '';
+  }
+  return String(data);
+}
+
+function mapValueToAstarteDataTuple(value: AstarteDataValue): AstarteDataTuple {
+  if (_.isNumber(value)) {
+    return { type: 'double', value };
+  } else if (_.isBoolean(value)) {
+    return { type: 'boolean', value };
+  } else if (_.isString(value)) {
+    return { type: 'string', value };
+  } else if (_.isArray(value)) {
+    if (value.length > 0) {
+      if (_.isNumber(value[0])) {
+        return { type: 'doublearray', value: value as number[] };
+      } else if (_.isBoolean(value[0])) {
+        return { type: 'booleanarray', value: value as boolean[] };
+      } else if (_.isString(value[0])) {
+        return { type: 'stringarray', value: value as string[] };
+      }
+    }
+    return { type: 'binaryblobarray', value: [] };
+  }
+  return { type: 'binaryblob', value: null };
+}
+
+const extractPublishedPaths = (interfaceValues: AstarteInterfaceValues): string[] => {
+  const paths: string[] = [];
+
+  if (
+    typeof interfaceValues === 'object' &&
+    interfaceValues !== null &&
+    !Array.isArray(interfaceValues)
+  ) {
+    const firstKey = Object.keys(interfaceValues)[0];
+
+    if (firstKey) {
+      const interfaceValue = interfaceValues[firstKey];
+
+      if (typeof interfaceValue === 'object' && interfaceValue !== null && !Array.isArray(interfaceValue)) {
+        const nestedInterfaceData = Object.keys(interfaceValue);
+
+        if (nestedInterfaceData.length === 1) {
+          const nestedKey = nestedInterfaceData[0];
+          paths.push(`/${firstKey}/${nestedKey}`);
+        } else if (nestedInterfaceData.length > 1) {
+          paths.push(`/${firstKey}`);
+        }
+      } else {
+        paths.push(`/${firstKey}`);
+      }
+    }
+  }
+
+  return paths;
+};
```

Comment:
This is still a bit convoluted, and I worry that it only works with a specific interface: i.e. if the interface mappings have a specific shape and nesting.
Let's double check that this works with other interfaces that have very different endpoints.

Besides that, maybe you could leave some explanatory comments to help the reader follow along, if simplifying the code is not possible.